### ISSUE TITLE
Shard count won't get translated properly

### ISSUE THREAD

**Issue by sjentzsch** [prometheus-operator/prometheus-operator#3801]
If I set `shards: 3` in my Prometheus CR, it will only spin up one additional shard:

```
monitoring       prometheus-kube-prometheus-stack-prometheus-0               3/3     Running   1          2m34s
monitoring       prometheus-kube-prometheus-stack-prometheus-shard-1-0       3/3     Running   1          2m34s
```
If I change shards back to 2, then back to 3 again, it will spin up the third one:

```
monitoring       prometheus-kube-prometheus-stack-prometheus-0               3/3     Running   1          7m46s
monitoring       prometheus-kube-prometheus-stack-prometheus-shard-1-0       3/3     Running   1          7m44s
monitoring       prometheus-kube-prometheus-stack-prometheus-shard-2-0       3/3     Running   1          4m7s
```
Looks to me like a controller issue?

**Environment**

* Prometheus Operator version: v0.44.1

* Kubernetes version information: 1.19.7

**Comment by lilic** [comment]
Thanks for the issue! How long did you wait for, as we do have a reconcile period, did it eventually (after 5-10 minutes) reconcile or did it never?

**Comment by sjentzsch** [comment]
I am afraid it does never, I waited 10min+
Only if I actively change the shards (3 -> 2 -> 3) it will add the third one.

**Comment by lilic** [comment]
Interesting this does indeed sound like a bug! Can you share some logs of prom-operator and the Prometheus instances that are created, e.g. the CRs? Hmm is this the helm chart and can you try to reproduce without the helm chart?

**Comment by Gokul-Radhakrishnan** [comment]
Even I am facing this issue (I did not use helm chart to install prometheus-operator). 

The following is the CR I used 

```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: prometheus
  labels:
    prometheus: `prometheus`
spec:
  serviceAccountName: prometheus
  serviceMonitorSelector:
    matchLabels:
      team: frontend
  shards: 5
  replicas: 2
```
The following are the logs from prometheus operator pod
```
level=info ts=2021-01-25T11:27:46.2592206Z caller=main.go:233 msg="Starting Prometheus Operator" version="(version=0.45.0, branch=refs/tags/v0.45.0, revision=5555f492df250168657b72bb8cb60bec071de71f)"
level=info ts=2021-01-25T11:27:46.2592909Z caller=main.go:234 build_context="(go=go1.15.6, user=lilic, date=20210113-16:18:28)"
level=info ts=2021-01-25T11:27:46.4481168Z caller=main.go:108 msg="Starting insecure server on [::]:8080"
level=info ts=2021-01-25T11:27:46.5478584Z caller=operator.go:300 component=thanosoperator msg="connection established" cluster-version=v1.19.2
level=info ts=2021-01-25T11:27:46.5480097Z caller=operator.go:309 component=thanosoperator msg="CRD API endpoints ready"
level=info ts=2021-01-25T11:27:46.5479535Z caller=operator.go:416 component=prometheusoperator msg="connection established" cluster-version=v1.19.2
level=info ts=2021-01-25T11:27:46.5481779Z caller=operator.go:425 component=prometheusoperator msg="CRD API endpoints ready"
level=info ts=2021-01-25T11:27:46.5487721Z caller=operator.go:438 component=alertmanageroperator msg="connection established" cluster-version=v1.19.2
level=info ts=2021-01-25T11:27:46.548941Z caller=operator.go:447 component=alertmanageroperator msg="CRD API endpoints ready"
level=info ts=2021-01-25T11:27:47.7464643Z caller=operator.go:261 component=thanosoperator msg="successfully synced all caches"
level=info ts=2021-01-25T11:27:48.0513965Z caller=operator.go:279 component=alertmanageroperator msg="successfully synced all caches"
level=info ts=2021-01-25T11:27:48.1503204Z caller=operator.go:355 component=prometheusoperator msg="successfully synced all caches"
level=info ts=2021-01-25T11:28:40.3154481Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:28:40.9794231Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:28:41.3907485Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:28:41.6105494Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:29:15.9845448Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:29:18.8104277Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:29:27.7280519Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:29:31.5418716Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:37:17.4242017Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:37:17.5965182Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:37:18.0979547Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:37:18.3595443Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:37:52.0250793Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:37:55.3104336Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:37:58.3696131Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
level=info ts=2021-01-25T11:38:08.0280193Z caller=operator.go:1161 component=prometheusoperator msg="sync prometheus" key=default/prometheus
```

I can see only 4 pods out of 10. (5 shards * 2 replica = 10 pods)
![image](https://user-images.githubusercontent.com/38651133/105702376-b8005d80-5f31-11eb-9a11-6177342bb995.png)

**Comment by lilic** [comment]
Thanks, that looks correct to me. Seems like we have a bug. If you want to @Gokul-Radhakrishnan or @sjentzsch either of you can have a look and fix it. Otherwise will mark as help wanted.

cc @brancz FYI.

**Comment by sidewinder12s** [comment]
Can also confirm we hit this as well setting shards to 5. Had to edit it back and forth to get it to add all the shards in.

**Comment by mikechengwei** [comment]
I can fix this pr. @lilic

**Comment by mikechengwei** [comment]
I found some log by enable debug log level.

```
level=debug ts=2021-04-11T16:10:13.24178926Z caller=operator.go:1190 component=prometheusoperator msg="reconciling statefulset" statefulset=prometheus-main shard=0
level=debug ts=2021-04-11T16:10:13.242372418Z caller=operator.go:1225 component=prometheusoperator msg="new statefulset generation inputs match current, skipping any actions"
``` 

After my analysis, I found a bug. When the first shard is created, it exits the loop. If the hash value of the first shard in the next loop has not changed, it will exit the loop directly and it will not create the next shard.

**Comment by simonpasquier** [comment]
closed by #3974
